File: src/test/java/net/sjrx/intellij/plugins/systemdunitfiles/annotators/LineContinuationCharacterFollowedByWhitespaceAnnotatorTest.java
package net.sjrx.intellij.plugins.systemdunitfiles.annotators;

import com.intellij.codeInsight.daemon.impl.HighlightInfo;
import com.intellij.codeInsight.daemon.impl.HighlightInfoType;
import net.sjrx.intellij.plugins.systemdunitfiles.AbstractUnitFileTest;

import java.util.List;

public class LineContinuationCharacterFollowedByWhitespaceAnnotatorTest extends AbstractUnitFileTest {
  
  public void testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedAsWarningInSimpleCase() {
    // Fixture Setup
    String file = "[Service]\n"
                  + "ExecStart=/bin/bash \\ \n"
                  + "X-Garbage=";
  
    setupFileInEditor("file.service", file);

    // Exercise SUT
    
    List<HighlightInfo> highlights = myFixture.doHighlighting();
    
  
    // Verification
    assertSize(1, highlights);
  
    HighlightInfo info = highlights.get(0);
    assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info.getDescription());
    assertEquals(HighlightInfoType.WARNING, info.type);
    
  
    assertEquals(" ", info.getText());
    
  }
  
  public void testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedAsWarningInLongerSimpleCase() {
    // Fixture Setup
    String file = "[Service]\n"
                  + "ExecStart=/bin/bash \\       \n"
                  + "X-Garbage=";
    
    setupFileInEditor("file.service", file);
    
    // Exercise SUT
    
    List<HighlightInfo> highlights = myFixture.doHighlighting();
    
    
    // Verification
    assertSize(1, highlights);
    
    HighlightInfo info = highlights.get(0);
    assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info.getDescription());
    assertEquals(HighlightInfoType.WARNING, info.type);
    
    
    assertEquals("       ", info.getText());
    
  }
  
  public void testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedOnASuccessfulLineContinuation() {
  
    // Fixture Setup
    String file = "[Service]\n"
                  + "ExecStart=/bin/bash \\\n"
                  + "X-Garbage=oeutnhouentuoeh\\ \t \n"
                  + "X-Garbage-Two=";
  
    setupFileInEditor("file.service", file);
  
    // Exercise SUT
  
    List<HighlightInfo> highlights = myFixture.doHighlighting();
  
  
    // Verification
    assertSize(1, highlights);
  
    HighlightInfo info = highlights.get(0);
    assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info.getDescription());
    assertEquals(HighlightInfoType.WARNING, info.type);
  
  
    assertEquals(" \t ", info.getText());
  
  }
  
  public void testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedOnASuccessfulLineContinuationWithComment() {
    // Fixture Setup
    String file = "[Service]\n"
                  + "ExecStart=/bin/bash \\\n"
                  + ";A nice comment\n"
                  + "X-Garbage=\\\t\t\t\n"
                  + "X-Garbage-Two=";
  
    setupFileInEditor("file.service", file);
  
    // Exercise SUT
  
    List<HighlightInfo> highlights = myFixture.doHighlighting();
  
  
    // Verification
    assertSize(1, highlights);
  
    HighlightInfo info = highlights.get(0);
    assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info.getDescription());
    assertEquals(HighlightInfoType.WARNING, info.type);
  
  
    assertEquals("\t\t\t", info.getText());
  }
  
  public void testThatNoExcessWhitespaceAfterLineContinuationCharacterResultsInNoWarnings() {
    // Fixture Setup
    String file = "[Service]\n"
                  + "ExecStart=/bin/bash \\\n"
                  + ";A nice comment\n"
                  + "X-Garbage=\\\n"
                  + "X-Garbage-Two=";
  
    setupFileInEditor("file.service", file);
  
    // Exercise SUT
  
    List<HighlightInfo> highlights = myFixture.doHighlighting();
  
  
    // Verification
    assertSize(0, highlights);
  }
  
  public void testThatNoLineContinuationCharacterResultsInNoWarnings() {
    // Fixture Setup
    String file = "[Service]\n"
                  + "ExecStart=/bin/bash \\\n"
                  + ";A nice comment\n"
                  + "X-Garbage=\t    \\\n"
                  + "X-Garbage-Two=";
  
    setupFileInEditor("file.service", file);
  
    // Exercise SUT
    List<HighlightInfo> highlights = myFixture.doHighlighting();
  
    // Verification
    assertSize(0, highlights);
  }
  
}

File: src/test/kotlin/net/sjrx/intellij/plugins/systemdunitfiles/annotators/LineContinuationCharacterFollowedByWhitespaceAnnotatorTest.kt
package net.sjrx.intellij.plugins.systemdunitfiles.annotators

import com.intellij.codeInsight.daemon.impl.HighlightInfoType
import junit.framework.TestCase
import net.sjrx.intellij.plugins.systemdunitfiles.AbstractUnitFileTest

class LineContinuationCharacterFollowedByWhitespaceAnnotatorTest : AbstractUnitFileTest() {
  fun testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedAsWarningInSimpleCase() {
    // Fixture Setup
    val file = ("[Service]\n"
      + "ExecStart=/bin/bash \\ \n"
      + "X-Garbage=")
    setupFileInEditor("file.service", file)

    // Exercise SUT
    val highlights = myFixture.doHighlighting()


    // Verification
    assertSize(1, highlights)
    val info = highlights[0]
    TestCase.assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info!!.description)
    TestCase.assertEquals(HighlightInfoType.WARNING, info.type)
    TestCase.assertEquals(" ", info.text)
  }

  fun testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedAsWarningInLongerSimpleCase() {
    // Fixture Setup
    val file = ("[Service]\n"
      + "ExecStart=/bin/bash \\       \n"
      + "X-Garbage=")
    setupFileInEditor("file.service", file)

    // Exercise SUT
    val highlights = myFixture.doHighlighting()


    // Verification
    assertSize(1, highlights)
    val info = highlights[0]
    TestCase.assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info!!.description)
    TestCase.assertEquals(HighlightInfoType.WARNING, info.type)
    TestCase.assertEquals("       ", info.text)
  }

  fun testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedOnASuccessfulLineContinuation() {

    // Fixture Setup
    val file = ("[Service]\n"
      + "ExecStart=/bin/bash \\\n"
      + "X-Garbage=oeutnhouentuoeh\\ \t \n"
      + "X-Garbage-Two=")
    setupFileInEditor("file.service", file)

    // Exercise SUT
    val highlights = myFixture.doHighlighting()


    // Verification
    assertSize(1, highlights)
    val info = highlights[0]
    TestCase.assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info!!.description)
    TestCase.assertEquals(HighlightInfoType.WARNING, info.type)
    TestCase.assertEquals(" \t ", info.text)
  }

  fun testThatExcessWhiteSpaceAfterLineContinuationCharacterIsDetectedOnASuccessfulLineContinuationWithComment() {
    // Fixture Setup
    val file = ("[Service]\n"
      + "ExecStart=/bin/bash \\\n"
      + ";A nice comment\n"
      + "X-Garbage=\\\t\t\t\n"
      + "X-Garbage-Two=")
    setupFileInEditor("file.service", file)

    // Exercise SUT
    val highlights = myFixture.doHighlighting()


    // Verification
    assertSize(1, highlights)
    val info = highlights[0]
    TestCase.assertEquals(LineContinuationCharacterFollowedByWhitespaceAnnotator.WARNING_MESSAGE, info!!.description)
    TestCase.assertEquals(HighlightInfoType.WARNING, info.type)
    TestCase.assertEquals("\t\t\t", info.text)
  }

  fun testThatNoExcessWhitespaceAfterLineContinuationCharacterResultsInNoWarnings() {
    // Fixture Setup
    val file = """
           [Service]
           ExecStart=/bin/bash \
           ;A nice comment
           X-Garbage=\
           X-Garbage-Two=
           """.trimIndent()
    setupFileInEditor("file.service", file)

    // Exercise SUT
    val highlights = myFixture.doHighlighting()


    // Verification
    assertSize(0, highlights)
  }

  fun testThatNoLineContinuationCharacterResultsInNoWarnings() {
    // Fixture Setup
    val file = ("[Service]\n"
      + "ExecStart=/bin/bash \\\n"
      + ";A nice comment\n"
      + "X-Garbage=\t    \\\n"
      + "X-Garbage-Two=")
    setupFileInEditor("file.service", file)

    // Exercise SUT
    val highlights = myFixture.doHighlighting()

    // Verification
    assertSize(0, highlights)
  }
}
